cmd/juju/application/bundlediff.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package application
+
+import (
+	"github.com/juju/bundlechanges"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"gopkg.in/juju/charm.v6"
+	"gopkg.in/juju/charmrepo.v3"
+	csparams "gopkg.in/juju/charmrepo.v3/csclient/params"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/api/annotations"
+	"github.com/juju/juju/api/application"
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/api/modelconfig"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/constraints"
+)
+
+const bundleDiffDoc = `
+Bundle can be a local bundle file or the name of a bundle in
+the charm store. The bundle can also be combined with overlays (in the
+same way as the deploy command) before comparing with the model.
+
+The map-machines option works similarly as for the deploy command, but
+existing is always assumed, so it doesn't need to be specified.
+
+Examples:
+    juju diff-bundle localbundle.yaml
+    juju diff-bundle canonical-kubernetes
+    juju diff-bundle mongodb-cluster --channel beta
+    juju diff-bundle canonical-kubernetes --overlay local-config.yaml --overlay extra.yaml
+    juju diff-bundle localbundle.yaml --map-machines 3=4
+
+See also:
+    deploy
+`
+
+// NewBundleDiffCommand returns a command to compare a bundle against
+// the selected model.
+func NewBundleDiffCommand() cmd.Command {
+	return modelcmd.Wrap(&bundleDiffCommand{})
+}
+
+// bundleDiffCommand compares a bundle to a model.
+type bundleDiffCommand struct {
+	modelcmd.ModelCommandBase
+	bundle         string
+	bundleOverlays []string
+	channel        csparams.Channel
+	annotations    bool
+
+	bundleMachines map[string]string
+	machineMap     string
+
+	// These are set in tests to enable mocking out the API and the
+	// charm store.
+	_apiRoot    base.APICallCloser
+	_charmStore BundleResolver
+}
+
+// IsSuperCommand is part of cmd.Command.
+func (c *bundleDiffCommand) IsSuperCommand() bool { return false }
+
+// AllowInterspersedFlags is part of cmd.Command.
+func (c *bundleDiffCommand) AllowInterspersedFlags() bool { return true }
+
+// Info is part of cmd.Command.
+func (c *bundleDiffCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "diff-bundle",
+		Args:    "<bundle file or name>",
+		Purpose: "Compare a bundle against a model and report any differences.",
```

Comment:
remove the trailing "."

Reply:
Removed.

Reply:
Gah, this is wrong - we have a test that enforces all Purposes must have a terminal "."!